zengo/service.py

```diff
@@ -0,0 +1,397 @@
+        Compare the User and ZendeskUser instances and determine whether we
+        need to update the data in Zendesk.
+
+        This method is suitable to be called when a local user changes their
+        email address or other user data.
+        """
+        changed = False
+        email_changed = False
+
+        if self.get_local_user_name() != remote_zd_user.name:
+            remote_zd_user.name = self.get_local_user_name(local_user)
+            changed = True
+
+        if self.get_local_user_external_id(local_user) != remote_zd_user.external_id:
+            remote_zd_user.external_id = self.get_local_user_external_id(local_user)
+            changed = True
+
+        if local_user.email and local_user.email != remote_zd_user.email:
+            remote_zd_user.email = local_user.email
+            changed = True
+            email_changed = True
+
+        if changed:
+            remote_zd_user = self.client.users.update(remote_zd_user)
+
+        if email_changed:
+            # then in addition to the above, we have to mark the newly
+            # created identity as verified and promote it to be primary
+            results = self.client.users.identities(id=remote_zd_user.id)
+            for identity in results:
+                if identity.value == local_user.email:
+                    self.client.users.identities.make_primary(
+                        user=remote_zd_user, identity=identity
+                    )
+                    break
+
+    def update_or_create_remote_zd_user(self, local_user):
+        remote_zd_user, is_definite_match = self.get_remote_zd_user_for_local_user(
+            local_user
+        )
+        if remote_zd_user and is_definite_match:
+            # check if we need to do any updates
+            self.update_remote_zd_user_for_local_user(local_user, remote_zd_user)
+        else:
+            remote_zd_user = self.create_remote_zd_user_for_local_user(local_user)
+        return remote_zd_user
+
+    def update_or_create_local_zd_user_for_remote_zd_user(self, remote_zd_user):
+        """
+        Given a RemoteZendeskUser instance, persist it as a LocalZendeskUser instance.
+        """
+        instance, created = LocalZendeskUser.objects.update_or_create(
+            zendesk_id=remote_zd_user.id,
+            defaults=dict(
+                # attempt to resolve the local user if possible
+                user=self.get_local_user_for_external_id(remote_zd_user.external_id),
+                email=remote_zd_user.email,
+                created_at=remote_zd_user.created_at,
+                name=remote_zd_user.name,
+            ),
+        )
+        return instance
+
+    @contextlib.contextmanager
+    def get_ticket_lock(self, ticket_id):
+        # detect both postgres and postgis, upon which we serialize ticket
+        # updates using a database lock
+        if "postg" in settings.DATABASES["default"]["ENGINE"]:
+            with advisory_lock("zengo-ticket-{}".format(ticket_id)):
```

Comment:
So from the docs, if `advisory_lock` keyword arg `wait` is `True` the default this could block, I am guessing indefinitely.

You can wait a specific amount of time and try, but the bigger question here is what happens if you can't get the lock within a reasonable period of time, or ever?

Reply:
The docs just say, "until" - so in the absence of further info I'd say you're right and that it waits indefinitely.

In practice, I've found these locks to be really robust and well behaved. But it might be too opinionated to do this here. It might be better left to the consuming application to decide to serialize updates if needed, based on the specific use-case. In production systems, the `process_event` action should probably be pushed onto a queue anyway, in which case this locked block ends up only serializing the queueing of the work, and not the work itself (hitting Zendesk APIs, firing signals, etc).

Reply:
[select for update](https://docs.djangoproject.com/en/2.1/ref/models/querysets/#select-for-update) could be typically be used for this kind of thing, but in the instance where no row exists yet for a Zendesk ticket, we have no row to lock on.